pkg/sql/plan/query_builder.go

```diff
@@ -954,15 +937,44 @@ func (builder *QueryBuilder) buildSelect(stmt *tree.Select, ctx *BindContext, is
 	}
 
 	var clause *tree.SelectClause
+	astOrderBy := stmt.OrderBy
+	astLimit := stmt.Limit
 
 	switch selectClause := stmt.Select.(type) {
 	case *tree.SelectClause:
 		clause = selectClause
 	case *tree.ParenSelect:
-		return builder.buildSelect(selectClause.Select, ctx, isRoot)
+		for {
+			if selectClause.Select.OrderBy != nil {
+				if astOrderBy != nil {
+					return 0, moerr.NewError(moerr.INVALID_INPUT, "multiple ORDER BY clauses not allowed")
+				}
+				astOrderBy = selectClause.Select.OrderBy
+			}
+			if selectClause.Select.Limit != nil {
+				if astLimit != nil {
+					return 0, moerr.NewError(moerr.INVALID_INPUT, "multiple LIMIT clauses not allowed")
+				}
+				astLimit = selectClause.Select.Limit
+			}
+
+			switch subClause := selectClause.Select.Select.(type) {
+			case *tree.SelectClause:
+				clause = subClause
+			case *tree.ParenSelect:
+				selectClause = selectClause.Select.Select.(*tree.ParenSelect)
+			case *tree.UnionClause:
+				return builder.buildUnion(subClause, astOrderBy, astLimit, ctx, isRoot)
+			default:
+				return 0, errors.New("", fmt.Sprintf("Statement '%s' will be supported in future version.", tree.String(stmt, dialect.MYSQL)))
+			}
+
+			if clause != nil {
+				break
+			}
+		}
 	case *tree.UnionClause:
-		return builder.buildUnion(stmt, ctx, isRoot)
-		// return 0, errors.New("", fmt.Sprintf("'%s' will be supported in future version.", selectClause.Type.String()))
+		return builder.buildUnion(selectClause, astOrderBy, astLimit, ctx, isRoot)
```

Comment:
This should work, but two `switch...case` are redundant.

First strip off any `*tree.ParenSelect`, then `switch...case` other types. This way it should be more readable.